crates/red_knot_python_semantic/src/types.rs

```diff
@@ -2056,26 +2056,58 @@ impl<'db> ClassType<'db> {
     /// Returns the inferred type of the class member named `name`.
     ///
     /// Returns [`Symbol::Unbound`] if `name` cannot be found in this class's scope
-    /// directly. Use [`ClassType::class_member`] if you require a method that will
+    /// directly. Use [`Class::class_member`] if you require a method that will
     /// traverse through the MRO until it finds the member.
     pub(crate) fn own_class_member(self, db: &'db dyn Db, name: &str) -> Symbol<'db> {
         let scope = self.body_scope(db);
         symbol(db, scope, name)
     }
 }
 
+/// A singleton type representing a single class object at runtime.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub struct ClassLiteralType<'db> {
+    class: Class<'db>,
+}
+
+impl<'db> ClassLiteralType<'db> {
+    fn member(self, db: &'db dyn Db, name: &str) -> Symbol<'db> {
+        self.class.class_member(db, name)
+    }
+
+    fn to_instance_type(self) -> InstanceType<'db> {
+        InstanceType::anonymous(self.class)
+    }
+}
+
+impl<'db> From<ClassLiteralType<'db>> for Type<'db> {
+    fn from(value: ClassLiteralType<'db>) -> Self {
+        Self::ClassLiteral(value)
+    }
+}
+
+/// A type representing the set of runtime objects which are instances of a certain class.
+///
+/// Some specific instances are important enough that they really constitute their own type
+/// which should compare and hash differently to other instances of the same class.
+/// Whether or not you're dealing with one of these symbols can be determined using the
+/// `known` field on the `InstanceType` struct.
+///
+/// Note that, for example, `InstanceType { class: typing._SpecialForm, known: None }`
+/// is a supertype of `InstanceType { class: typing._SpecialForm, known: KnownInstance::Literal }`.
+/// The two types are not disjoint.
```

Comment:
I'm glad I tried to writeup some docs here, as it made me realise that `InstanceType` types can now represent two pretty different things, depending on whether the `known` field is `Some()` or `None`. I'm not sure that's really desirable; we might consider having a separate `Type::KnownInstance` variant on `Type`.

Reply:
I agree. I also realized that we are now handling known instances wrongly at least in `Type::is_equivalent_to` and likely in other `Type` methods as well.

Although it was convenient to be able to ignore the `known` field when unpacking a `Type::Instance`, I think it makes it too easy to do the wrong thing.

Reply:
https://github.com/astral-sh/ruff/issues/14114